src/runtime/runtime_nintendoswitch.s

```diff
@@ -0,0 +1,54 @@
+.section .text.armGetSystemTick, "ax", %progbits
+.global  armGetSystemTick
+.type    armGetSystemTick, %function
+.align 2
+armGetSystemTick:
+    mrs x0, cntpct_el0
+    ret
+
+.section .text.armGetSystemTickFreq, "ax", %progbits
+.global  armGetSystemTickFreq
+.type    armGetSystemTickFreq, %function
+.align 2
+armGetSystemTickFreq:
+    mrs x0, cntfrq_el0
+    ret
+
+.section .text.nxOutputString, "ax", %progbits
+.global nxOutputString
+.type nxOutputString, %function
+.align 2
+.cfi_startproc
+nxOutputString:
+	svc 0x27
```

Comment:
You could look at `arm.SVCall2` for these `svc` wrappers. It has the benefit of avoiding assembly.

Reply:
I tried to use `arm.SVCall2` but it uses `r0` which is 32 bit, and the aarch64 requires a 64 bit register `x0`.

If I go to `compiler/inlineasm.go` and change the function `emitSVCall(args []ssa.Value)`:

```go
constraints := "={r0}"
```
to 
```go
constraints := "={x0}"
```

and 

```
constraints += ",{r" + strconv.Itoa(i) + "}"
```

to 

```
constraints += ",{x" + strconv.Itoa(i) + "}"
```

it does work.

So to avoid changing that inline asm code and not having a .s file I'm doing this:

```go
func NxOutputString(str *byte, size uint64) uint64 {
    res := arm.AsmFull(
        `mov x0, {str}
         mov x1, {size}
         svc 0x27`,
        map[string]interface{}{
            "str":  uintptr(unsafe.Pointer(str)),
            "size": size,
        })
	return uint64(res)
}
```

Is that better?

Reply:
That makes more sense even to me. @aykevl seems legit?

Reply:
I think we should fix `emitSVCall` eventually, I guess this is fine for now

Reply:
Adding support for aarch64 should be relatively easy. @racerxdl do you have any information on the clobbered/retained registers after a `svc` instruction? On arm it's the usual calling convention hence the clobbered registers.
The inline assembly you posted above may not be correct because of clobbered registers, so if you don't know it may be best to assume the usual calling convention (which will be used when you call an assembly function).

Reply:
Hi @aykevl , yes they're all mapped here: https://switchbrew.org/wiki/SVC

Which in the case of this SVC is this: https://switchbrew.org/wiki/SVC#OutputDebugString

![image](https://user-images.githubusercontent.com/578310/87707315-4c198d00-c777-11ea-8d15-b907b337b9c4.png)

Reply:
That's not what I mean. Often, such syscall instructions (`svc` is very similar to the amd64 `syscall` instruction) clobber a number of registers, possibly including their input registers. This means the registers many have any value. For a function call to an unknown function (such as assembly) the compiler knows that some registers will be clobbered and won't assume the values in the registers remain the same. For regular inline assembly,  the compiler is not aware of this and it may (or may not) lead to corruption at runtime.

Reply:
The `arm.SVCall` functions are implemented in a way to be aware of this: they clobber `r1`, `r2` and `r3` following the ARM calling convention (`r0` is an output register and thus is implicitly clobbered). Inline assembly does not automatically mark these registers as clobbered.

Reply:
Ahhh, so this function just inlines whatever I put in that assembly statement inside the function its being called?

 Well I'm not sure to be honest. The SVC calls in both libnx and libtransistor are done in pure assembly but called from C (they're exported as normal functions).

You think its safer to keep then in a `.s` file just to avoid register corruption?

Reply:
@aykevl what do you think?

Reply:
@deadprogram / @aykevl I think that's the only missing point now.

Reply:
Ok reverted to use assembly code instead.